src/zcash/NoteEncryption.cpp

```diff
@@ -0,0 +1,169 @@
+#include "NoteEncryption.hpp"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+
+#define NOTEENCRYPTION_CIPHER_KEYSIZE 32
+
+void clamp_curve25519(unsigned char key[crypto_scalarmult_SCALARBYTES])
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &hSig,
+    unsigned char nonce
+   )
+{
+    if (nonce == 0xff) {
+        throw std::logic_error("no additional nonce space for KDF");
+    }
```

Comment:
You could use an `unsigned int` instead of an `unsigned char` and then check `> 0xff` but then you'd need to copy it into an `unsigned char` variable below. I think it's fine the way it is.